### ISSUE TITLE
AsyncAPI parameters examples (was URI_Param test to response)

### ISSUE THREAD

**Issue by Eyolon** [microcks/microcks#727]
Hi
I try to use parameterized uri websocket subscribe endpoint. Something like this in my asyncApi
```
channels:
  'camera/{cameraId}':
      description: Use it to retrieve info about camera by id
      parameters:
        cameraId:
          $ref: '#/components/parameters/cameraId'
      bindings:
        ws:
          bindingVersion: 0.1.0
      publish:
        message:
          $ref: '#/components/messages/cameraMessage'
        
#[...]many lines later 

 parameters:
    cameraId:
      description: The ID of camera.
      schema:
        type: string
        enum: ['1']

#[...] again

components:
  messages:
    cameraMessage:
          name: camera
          title: camera object
          summary: camera answer
          payload:
            $ref: '#/components/schemas/camera'
          examples:
            - name: cameraObject
              summary: camera 
              payload:
                cameraId: 1
                name: "Camera 1"
```

doc generation ok but in microcks (docker using, so in "latest" release), it generate:

`localhost:8081/api/ws/myApi/1.0.0/camera/undefined`

and in logs of async minion i see : 

```
2022-12-28 14:35:11,836 INFO  [io.git.mic.min.asy.pro.WebSocketProducerManager] (QuarkusQuartzScheduler_Worker-12) Publishing on channel {/api/ws/myApi/1.0.0/camera/{cameraId}}, message: {"cameraId":1,"name":"Camera 1"} 
```
I've try to open my test on :

* ws://localhost:8081/api/ws/myApi/camera/undefined -> doesn't work
* ws://localhost:8081/api/ws/myApi/camera/{cameraId} -> doesn't work
* ws://localhost:8081/api/ws/myApi/camera/%7BcameraId%7D -> doesn't work
* ws://localhost:8081/api/ws/myApi/camera/1-> doesn't work
* ws://localhost:8081/api/ws/myApi/camera/-> doesn't work

In case of it's not a bug and i'm just tired by years end holidays, where are documentation ? in other way, maybe looking for a possibilitie to use URI_PARTS dispatcher to match a special value that answer an available message [(like JSON BODY in this docs)](https://microcks.io/documentation/using/advanced/dispatching/)

Thanks for your help !

**Comment by Eyolon** [comment]
Ok after one more test with x-microcks-operation i've some clues on the problem :

i've add an x-microcks-operation in my asyncApi

```
channels:
  'camera/{cameraId}':
      description: Use it to retrieve info about camera by id
      parameters:
        cameraId:
          $ref: '#/components/parameters/cameraId'
      bindings:
        ws:
          bindingVersion: 0.1.0
      publish:
        x-microcks-operation:
          delay: 3
          dispatcher: SCRIPT
          dispatcherRules: |
            def path = mockRequest.getRequest().getRequestURI();
            if (path.contains("camera/1")) {
              return "cameraMessage";
            }
            return "FAIL";
      message:
        $ref: '#/components/messages/cameraMessage'
```
now it display that my endpoint is on `localhost:8081/api/ws/myApi/camera/{cameraId}` and my dispatcher is as well SCRIPT and not URI_PARTS

BUT even if i connect on 
* ws://localhost:8081/api/ws/myApi/camera/1 -> doesn't work (can connect like previously but no response with message )
* ws://localhost:8081/api/ws/myApi/camera/{cameraId} -> doesn't work (can connect like previously but no "FAIL" response)

So just to be sure, i've test on a worked endpoint subscribtion (case of the problem came from connection ?) and with the same script, never received "FAIL" but the expected worked message (answer without x-microcks-operation)

So maybe dispatcher in asyncApi simply doesnt works ?

**Comment by lbroudoux** [comment]
Hi @Eyolon 
Thanks for the issue and sorry that it - apparently - take you some time to test and figured it out...

First thing here is that `dispatcher` are - for the moment - not used when dealing with asynchronous API. The mock engine published all mock messages on various endpoints at defined frequency. That's it. No `dispatcher` involve, no routing rules, ...

Then, your above AsyncAPI contract is just missing one thing to be able to publish on `ws://localhost:8081/api/ws/myApi/camera/1` : 1 example value for `cameraId` parameter. Documentation is not present at the moment for that point because actual implementation is a bit cranky and we expect enhancing things in coming weeks (see #600)

As of today, to get it running you have to add `examples` at the schema level for your parameter. Below a complete AsyncAPI specification that handles this case:

```yaml
asyncapi: '2.3.0'
info:
  title: myAPI
  version: 1.0.0
defaultContentType: application/json
channels:
  'camera/{cameraId}':
    description: Use it to retrieve info about camera by id
    parameters:
      cameraId:
        $ref: '#/components/parameters/cameraId'
    bindings:
      ws:
        bindingVersion: 0.1.0
    publish:
      message:
        $ref: '#/components/messages/cameraMessage'
components:
  parameters:
    cameraId:
      description: The ID of camera.
      schema:
        type: string
        enum: [ '1' ]
        examples:
          cameraObject:
            value: '1'
  messages:
    cameraMessage:
      name: camera
      title: camera object
      summary: camera answer
      payload:
        $ref: '#/components/schemas/camera'
      examples:
        - name: cameraObject
          summary: camera
          payload:
            cameraId: "1"
            name: "Camera 1"
  schemas:
    camera:
      type: object
      required:
        - name
        - cameraId
      properties:
        name:
          type: string
        cameraId:
          type: string
```

Hope it helps!

**Comment by lbroudoux** [comment]
Will be fixed with #600

**Comment by Eyolon** [comment]
Hi
Yeah with your explanation it works finaly as expected. But the syntax apears invalid. parameter.xxxx.schema.examples must be an array and in your solution and #600 or [your example](https://github.com/microcks/microcks/blob/master/webapp/src/test/resources/io/github/microcks/util/asyncapi/api-maintenance.async-api-spec-gh-master.yaml) it's an object. Not a big problem for me at this moment but you cant generate code with this syntax.

Thank you for your precious help anyway ;)
